File: simple_sample/rb/generator.rb
# 1) Generator 相当：Enumerator
def generator(max)
  Enumerator.new do |y|
    puts "  start generator...."
    a = 1
    b = 1
    while b < max
      puts "   start yield #{b}"
      cmd = y.yield(b)
      puts "   end   yield #{b} #{cmd}"

      case cmd
      when "Stop"
        # JS の g.return('ok') っぽく：ここで終了値 "ok"
        puts "   stop by outer"
        raise StopIteration, "Stopped"
      end
      a, b = b, a + b
    end
    puts "  end loop...."
    # 自然終了時に終了値を付けたいなら（任意）
  end
end

puts "== Generator demo =="
puts '開始'
g = generator(100)
puts "generator: #{g.inspect}"
g.each { |i| puts "呼び出し元： #{i} #{g.inspect}" }

# Rubyでも呼び出し元からジェネレータに値を設定できる
puts '------------------'
g2 = generator(100)
next1 = g2.next
puts "next -> #{next1} (#{g2.inspect})"
puts "feed...."
g2.feed "Stop"
begin
  g2.next              # ここで :stop が反映されて終了
rescue StopIteration => e
  p e
end

File: simple_sample/node/generator.js
// https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Global_Objects/Generator

function *generator(max) {
  console.log('  start generator....')
  let a = 1;
  let b = 1;
  try {
    while (max > b) {
      console.log("   start yield", b);
      yield b;
      console.log("   end yield", b);
      a = b;
      b = a + b;
    }
  } catch (ex) {
    console.error(ex)
    return "xxxx";
  }
}

console.log('開始');
const g = generator(100);
console.log("generator:", g);
for (const i of g) {
  console.log("呼び出し元：", i, g)
}

console.log('------------------')
const g2 = generator(100);
console.log(g2.next(), g2);
console.log(g2.return('ok'));

console.log('------------------')
const g3 = generator(100);
console.log(g3.next(), g2);
console.log(g3.throw('error'));
